### PR TITLE
Compressed/combined OTA updates

### PR DESCRIPTION
**Pull request by sergeuz**
### Problem

This PR implements support for compressed/combined OTA updates on Gen 3 platforms.

Notes:

- A bootloader update is required.
- For now, only system and user part modules can be compressed/combined.
- This feature requires server-side support. We haven't yet deployed it.
- Unit tests are currently failing on CI due to missing zlib libraries. An update to the buildpack is pending.

### Steps to Test

1. Change the bootloader version in system_module_version.mk to 1000 (2.0.0), compile and flash the bootloader to a Gen 3 device.
2. Change the system firmware version to 1502 (1.5.0), compile and flash the system part and example app to the device.
3. Change the system firmware version to 1511 (1.5.1), recompile the app and flash it to the device OTA.
4. Verify via the log that the device received a compressed binary. Device should update from 1.5.0 to 1.5.1 without entering safe mode.

### Example App

```cpp
#include "application.h"

SYSTEM_MODE(SEMI_AUTOMATIC)
// SYSTEM_THREAD(ENABLED)

namespace {

system_tick_t updateStart = 0;

void firmwareUpdateEvent(system_event_t event, int param, void* data) {
    switch (param) {
    case firmware_update_begin: {
        const auto file = (FileTransfer::Descriptor*)data;
        Log.info("Firmware update started; file size: %uK", (unsigned)file->file_length / 1024);
        updateStart = millis();
        break;
    }
    case firmware_update_complete: {
        Log.info("Firmware update completed in %us", (unsigned)(millis() - updateStart) / 1000);
        break;
    }
    case firmware_update_failed: {
        Log.error("Firmware update failed");
        break;
    }
    default:
        break;
    }
}

void resetPendingEvent(system_event_t event, int param, void* data) {
    HAL_Delay_Milliseconds(1000);
}

const Serial1LogHandler logHandler(115200, LOG_LEVEL_WARN, {
    { "comm.ota", LOG_LEVEL_ALL },
    { "hal.ota", LOG_LEVEL_ALL },
    { "app", LOG_LEVEL_ALL }
});

} // namespace

void setup() {
    System.on(firmware_update, firmwareUpdateEvent);
    System.on(reset_pending, resetPendingEvent);
    Particle.connect();
}

void loop() {
}
```

### References

- [ch53174]

### CODE REVIEW

**Comment by avtolstoy** [comment]
I've also verified that bootloader, SoftDevice and NCP updates work correctly.

**Comment by avtolstoy** [comment]
I'm going to merge this PR today breaking develop and will fix the zlib issue in a separate PR along with the compiler change.

File: hal/src/nRF52840/bootloader.h [hunk 1]
```diff
@@ -19,6 +19,7 @@ bool bootloader_update_if_needed();
 
 int bootloader_update(const void* bootloader_image, unsigned length);
 
+uint16_t get_bootloader_version(void);
```
[the remark this answers is not in the document]
Comment (sergeuz) [reply hal/src/nRF52840/bootloader.h:22]: hm, it turns out we already have `Bootloader_Get_Version()` in the platform code, but it's unused and works via system flags, which is a less reliable approach. I'll rename this function to `bootloader_get_version()` and add a FIXME to figure out this duplication later

[the remark this answers is not in the document]
Comment (sergeuz) [reply hal/src/nRF52840/bootloader.h]: "Fixed" :+1:

**Pull request merged**